### PR TITLE
do not crash readXLMetaNoData - if the `xl.meta` has incorrect content

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
do not crash readXLMetaNoData - if the `xl.meta` has incorrect content

## Motivation and Context
```
tmp = buf[want:]
```

Would potentially crash when `buf` is truncated for some reason
and does not have the expected bytes, this is of course considered
not normal and is an odd situation. But we do not attempt to crash
here instead allow for errors to be returned.

## How to test this PR?
Following attached xl.meta file will crash the server. 
[xl.meta.gz](https://github.com/minio/minio/files/8238768/xl.meta.gz)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/xl-storage-format-v2.go [hunk 1]
```diff
@@ -699,7 +699,7 @@ func readXLMetaNoData(r io.Reader, size int64) ([]byte, error) {
 		case 1, 2, 3:
 			sz, tmp, err := msgp.ReadBytesHeader(tmp)
 			if err != nil {
-				return nil, err
+				return nil, fmt.Errorf("readXLMetaNoData(msgp.ReadBytesHeader): uknown metadata version %w -> %s ", err, path)
```
Comment (klauspost) [inline cmd/xl-storage-format-v2.go]:
```suggestion
				return nil, fmt.Errorf("readXLMetaNoData(read_meta): %w -> %s ", err, path)
```

File: cmd/xl-storage-format-v2.go [hunk 2]
```diff
@@ -720,20 +720,24 @@ func readXLMetaNoData(r io.Reader, size int64) ([]byte, error) {
...
 
+			if int64(len(buf)) < want {
+				return nil, fmt.Errorf("buffer shorter than expected buflen: %d, want: %d -> %s", len(buf), want, path)
+			}
+
 			tmp = buf[want:]
 			_, after, err := msgp.ReadUint32Bytes(tmp)
 			if err != nil {
-				return nil, err
+				return nil, fmt.Errorf("readXLMetaNoData(msgp.ReadUint32Bytes): unknown metadata version %w -> %s ", err, path)
```
Comment (klauspost) [inline cmd/xl-storage-format-v2.go]:
```suggestion
				return nil, fmt.Errorf("readXLMetaNoData(read_crc): %w -> %s" , err, path)
```

File: cmd/xl-storage-format-v2.go [hunk 3]
```diff
@@ -647,7 +647,7 @@ func metaDataPoolPut(buf []byte) {
 // readXLMetaNoData will load the metadata, but skip data segments.
 // This should only be used when data is never interesting.
 // If data is not xlv2, it is returned in full.
-func readXLMetaNoData(r io.Reader, size int64) ([]byte, error) {
+func readXLMetaNoData(r io.Reader, size int64, path string) ([]byte, error) {
```
Comment (klauspost) [inline cmd/xl-storage-format-v2.go]: Add the path to the error upstream.

File: cmd/xl-storage-format-v2.go [hunk 4]
```diff
@@ -720,20 +720,24 @@ func readXLMetaNoData(r io.Reader, size int64) ([]byte, error) {
 				return nil, err
 			}
 
+			if int64(len(buf)) < want {
+				return nil, fmt.Errorf("buffer shorter than expected buflen: %d, want: %d -> %s", len(buf), want, path)
```
Comment (klauspost) [inline cmd/xl-storage-format-v2.go]: Maybe wrap an errCorrupt

**Review by klauspost — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by klauspost — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**